cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/SortedReplicasHelper.java

```diff
@@ -0,0 +1,151 @@
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+
+/**
+ * A helper class to create new {@link SortedReplicas} for cluster/broker/disk.
+ * This class provides fluent API to enable user to customize {@link SortedReplicas} by adding new selection/priority/score
+ * functions one by one.
+ *
+ * One example is as follows( to register a SortedReplica to clusterModel which includes only follower replicas, prioritize offline
+ * replicas and sort replicas based on replica'disk utilization).
+ * <p>
+ * <pre>
+ * {@code
+ * new SortedReplicasHelper().addSelectionFunc(ReplicaSortFunctionFactory.selectFollowers())
+ *                           .addPriorityFunc(ReplicaSortFunctionFactory.prioritizeOfflineReplicas());
+ *                           .addScoreFunc(addScoreFunc(ReplicaSortFunctionFactory.sortByMetricGroupValue(DISK.name())))
+ *                           .trackSortedReplicasFor(sortName, clusterModel)
+ * } </pre>
+ * </p>
+ */
+public class SortedReplicasHelper {
+
+  private Set<Function<Replica, Boolean>> _selectionFuncs;
+  private List<Function<Replica, Integer>> _priorityFuncs;
+  private Function<Replica, Double> _scoreFunc;
+
+  public SortedReplicasHelper() {
+    _selectionFuncs = new HashSet<>();
+    _priorityFuncs = new ArrayList<>();
+  }
+
+  /**
+   * If condition hold, add a selection function to the {@link SortedReplicas} to be created.
+   *
+   * @param selectionFunc The selection function to add.
+   * @param condition Whether add the selection function or not.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper maybeAddSelectionFunc(Function<Replica, Boolean> selectionFunc, boolean condition) {
+    if (condition) {
+      _selectionFuncs.add(selectionFunc);
+    }
+    return this;
+  }
+
+  /**
+   * Add a selection function to the {@link SortedReplicas} to be created.
+   *
+   * @param selectionFunc The selection function to add.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper addSelectionFunc(Function<Replica, Boolean> selectionFunc) {
+    return maybeAddSelectionFunc(selectionFunc, true);
+  }
+
+
+  /**
+   * If condition hold, add a priority function to the {@link SortedReplicas} to be created.
+   * Note the order adding priority function matters. Each call of this method or {@link this#addPriorityFunc(Function)}
+   * will add new priority function to the end of priority function list, which will be applied only when all the previous
+   * priority functions cannot resolve two replica's priority.
+   *
+   * @param priorityFunc The priority function to add.
+   * @param condition Whether add the priority function or not.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper maybeAddPriorityFunc(Function<Replica, Integer> priorityFunc, boolean condition) {
+    if (condition && !_priorityFuncs.contains(priorityFunc)) {
```

Comment:
If the intention is to ensure that the `priorityFunc` is inserted just once, is there another more suitable data structure we can use (e.g. `SortedSet`)?

Reply:
Changed to LinkedHashSet since here we want to preserve the insertion order.